### ISSUE TITLE
slave mode need masterauth

### ISSUE THREAD

**Issue by shanyou** [yinqiwen/ardb#179]
I am using taobao KV store service, it provides Redis compatible protocal. Now I want to use ardb slave mode to backup the database. but it need password for auth. I have searched the source code of ardb replication, but it not support the password authorization.  Can you fix this issue or tell me how to setup develop environment and how to debug. so I can fixed it myself.

**Comment by yinqiwen** [comment]
All the code of slave is in ' replication/slave.cpp'
you can do that when slave connected master from  https://github.com/yinqiwen/ardb/blob/master/src/replication/slave.cpp#L92

**Comment by shanyou** [comment]
After many studies, taobao kv store not support SYNC command yet, so it not possible to sync data by slave mode. I have to find another way backup it. Thank you for your kind reply.

**Comment by pedigree** [comment]
While taobao KV doesnt support it yet, many other servers do.  Azure for example mandates AUTH and ardb is currently non functional as a slave when attempting to connect to an azure redis master.

**Comment by pedigree** [comment]
Please reopen this shanyou as it should be included.

**Comment by pedigree** [comment]
A heads up, Azure Redis does NOT provide persistence so if you want to make sure you dont lose data, ardb would be a perfect solution to keep a on disk copy.. if masterauth worked :(

**Comment by pedigree** [comment]
This adds "masterauth password" to ardb.conf and I currently have this backing up an azure redis instance.  

It's *not* heavily tested and you must "make clean" or you get random segfaults.

```
diff --git a/src/ardb.cpp b/src/ardb.cpp
index 57030ee..0d0e7fa 100755
--- a/src/ardb.cpp
+++ b/src/ardb.cpp
@@ -917,7 +917,7 @@ OP_NAMESPACE_BEGIN
         RedisCommandHandlerSetting* found = FindRedisCommandHandlerSetting(args);
         if (NULL == found)
         {
-            ERROR_LOG("No handler found for:%s with size:%u in connection:%u", args.GetCommand().c_str(), args.GetCommand().size(), ctx.client->GetID());
+            ERROR_LOG("No handler found for:%s with size:%u in connection:%u", args.GetCommand().c_str(), args.GetCommand().size(), ctx.client ? ctx.client->GetID() : 0xffff);
             ERROR_LOG("Invalid command's ascii codes:%s", ascii_codes(args.GetCommand()).c_str());
             fill_error_reply(ctx.reply, "unknown command '%s'", args.GetCommand().c_str());
             ctx.close_after_processed = true;
diff --git a/src/config.cpp b/src/config.cpp
index 57a2f42..b46bac9 100755
--- a/src/config.cpp
+++ b/src/config.cpp
@@ -306,6 +306,8 @@ OP_NAMESPACE_BEGIN
 
         conf_get_bool(props, "lua-exec-atomic", lua_exec_atomic);
 
+        conf_get_string(props, "masterauth", master_auth);
+
         trusted_ip.clear();
         Properties::const_iterator ip_it = props.find("trusted-ip");
         if (ip_it != props.end())
diff --git a/src/config.hpp b/src/config.hpp
index 3ff51fc..81fefd0 100755
--- a/src/config.hpp
+++ b/src/config.hpp
@@ -159,6 +159,8 @@ OP_NAMESPACE_BEGIN
 
             bool lua_exec_atomic;
 
+            std::string master_auth;
+
             ArdbConfig() :
                     daemonize(false), unixsocketperm(755), max_clients(10000), tcp_keepalive(0), timeout(0), slowlog_log_slower_than(
                             10000), slowlog_max_len(128), repl_data_dir("./repl"), backup_dir("./backup"), backup_redis_format(
@@ -177,7 +179,7 @@ OP_NAMESPACE_BEGIN
                             5000), primary_port(0), slave_client_output_buffer_limit(256 * 1024 * 1024), pubsub_client_output_buffer_limit(
                             32 * 1024 * 1024), slave_ignore_expire(false), slave_ignore_del(false), repl_disable_tcp_nodelay(
                             false), scan_redis_compatible(true), scan_cursor_expire_after(60), max_string_bitset_value(
-                            1024 * 1024), maxdb(16),lua_exec_atomic(true)
+                            1024 * 1024), maxdb(16),lua_exec_atomic(true), master_auth("")
             {
             }
             bool Parse(const Properties& props);
diff --git a/src/replication/slave.cpp b/src/replication/slave.cpp
index b7c4d64..176a0fe 100755
--- a/src/replication/slave.cpp
+++ b/src/replication/slave.cpp
@@ -45,7 +45,7 @@ namespace ardb
             SLAVE_STATE_CLOSED), m_cron_inited(false), m_cmd_recved_time(0), m_master_link_down_time(0), m_server_type(
             ARDB_DB_SERVER_TYPE), m_server_support_psync(false), m_actx(
             NULL), m_rdb(NULL), m_backlog(serv->m_repl_backlog), m_routine_ts(0), m_cached_master_repl_offset(0), m_cached_master_repl_cksm(
-                    0), m_lastinteraction(0)
+                    0), m_lastinteraction(0), m_sent_auth_info(false)
     {
     }
 
@@ -92,9 +92,20 @@ namespace ardb
     void Slave::ChannelConnected(ChannelHandlerContext& ctx, ChannelStateEvent& e)
     {
         DEBUG_LOG("Master conn connected.");
+
+        if (m_serv->m_cfg.master_auth.length() > 0)
+        {
+            Buffer auth;
+            auth.Printf("auth %s\r\n", m_serv->m_cfg.master_auth.c_str());
+            INFO_LOG("[Slave]Send authentication command");
+            m_sent_auth_info = true;
+            ctx.GetChannel()->Write(auth);
+        }
+
         Buffer info;
         info.Printf("info Server\r\n");
         ctx.GetChannel()->Write(info);
+
         m_slave_state = SLAVE_STATE_WAITING_INFO_REPLY;
         m_lastinteraction = m_cmd_recved_time = time(NULL);
         m_master_link_down_time = 0;
@@ -175,6 +186,19 @@ namespace ardb
     }
     void Slave::HandleRedisReply(Channel* ch, RedisReply& reply)
     {
+printf("Slave::HandleRedisReply  type:%d  state:%d  m_sent_auth_info:%d\n", reply.type, m_slave_state, m_sent_auth_info ? 1 : 0);
+        if (m_sent_auth_info)
+        {
+            m_sent_auth_info = false;
+
+            if (reply.type == REDIS_REPLY_STATUS)
+                return;
+
+            ERROR_LOG("Recv auth reply error:%s", reply.str.c_str());
+            Close();
+            return;
+        }
+
         switch (m_slave_state)
         {
             case SLAVE_STATE_WAITING_INFO_REPLY:
diff --git a/src/replication/slave.hpp b/src/replication/slave.hpp
index 643ee80..683f429 100755
--- a/src/replication/slave.hpp
+++ b/src/replication/slave.hpp
@@ -87,6 +87,8 @@ namespace ardb
 
             time_t m_lastinteraction;
 
+            bool m_sent_auth_info;
+
             void HandleRedisCommand(Channel* ch, RedisCommandFrame& cmd);
             void HandleRedisReply(Channel* ch, RedisReply& reply);
             void HandleRedisDumpChunk(Channel* ch, RedisDumpFileChunk& chunk);
```

**Comment by yinqiwen** [comment]
fixed in 0.9

**Comment by muyinliu** [comment]
Still not fix in 0.9 . I pull a request #236, please check.
